substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -660,7 +660,7 @@ public String name() {
 
             @Override
             public String valueType() {
-                return hostedMethod.getSignature().getReturnType(null).toJavaName();
+                return toJavaName((HostedType) hostedMethod.getSignature().getReturnType(null));
```

Comment:
@adinn @zakkak This fixes some errors related to Lambda types. `LambdaSubstitutionType` injects a new stable name for such types. For the CE debug information, you unwrap the substitutions (I don't really know why, but certainly don't want to change it either). These two usages were the only ones I could find that were missing the unwrapping.

Reply:
@christianwimmer thanks for catching and fixing this. Your changes look good to me and I was also not able to find other usages missing the unwrapping.

I don't really get how this change relates to `LambdaSubstitutionType`'s stableName though. AFAICS `LambdaSubstitutionType#toJavaName()` invokes `toJavaName()` on the original `ResolvedJavaType`, so it's now using the `stableName`. https://github.com/oracle/graal/blob/5882aaa3e2aef4dc2ff5428751ad7d942af9373c/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/lambda/LambdaSubstitutionType.java#L310-L313

As for the why we do the unwrapping:

https://github.com/oracle/graal/blob/5882aaa3e2aef4dc2ff5428751ad7d942af9373c/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java#L177-L187

Reply:
With my current work to use more factory methods it is now possible that a lambda type is used as the return type of a method (that cannot occur in "normal" bytecode because the lambda type is only created on-the-fly by the `LambdaMetaFactory` and not mentioned as a return type in any generated method).

With a lambda type as a method return type, some code paths were using the class name generated by the `LambdaMetaFactory`, and some code paths were using the new name as rewritten by `LambdaSubstitutionType`. Your debug info code goes back to the original name from HotSpot, so it is now (consistently) using the original name that comes from `LambdaMetaFactory`.